lib/fileset/internal.nix

```
@@ -894,7 +901,9 @@ rec {
           # However a simpler alternative still would be [a builtins.gitLsFiles](https://github.com/NixOS/nix/issues/2944).
           fetchResult = fetchGit ({
             url = path;
-          } // extraFetchGitAttrs);
+          }
+          // (optionalAttrs (versionAtLeast nixVersion _fetchGitShallowMinver) { shallow = true; })
```

```suggestion
          # In older Nix versions, repositories were always assumed to be deep clones, which made `fetchGit` fail for shallow clones
          # For newer versions this was fixed, but the `shallow` flag is required.
          # The only behavioral difference is that for shallow clones, `fetchGit` doesn't return a `revCount`,
          # which we don't need here, so it's fine to always pass it.
          
          # Unfortunately this means older Nix versions get a poor error message for shallow repositories, and there's no good way to improve that.
          # Checking for `.git/shallow` doesn't seem worth it, especially since that's more of an implementation detail,
          # and would also require more code to handle worktrees where `.git` is a file.
          // optionalAttrs (versionAtLeast nixVersion _fetchGitShallowMinver) { shallow = true; }
```

In addition to the long comment, I also removed some parenthesis that weren't necessary.